ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java

```diff
@@ -230,31 +231,39 @@ private List<Withdrawal> getExpectedWithdrawals(final BeaconStateCapella preStat
     final List<Withdrawal> expectedWithdrawals = new ArrayList<>();
     final WithdrawalSchema withdrawalSchema = schemaDefinitionsCapella.getWithdrawalSchema();
     final UInt64 epoch = miscHelpers.computeEpochAtSlot(preState.getSlot());
-    final int validatorCount = preState.getValidators().size();
+    final SszList<Validator> validators = preState.getValidators();
+    final SszUInt64List balances = preState.getBalances();
+    final int validatorCount = validators.size();
+    final int maxValidatorsPerPayload = specConfigCapella.getMaxWithdrawalsPerPayload();
     UInt64 withdrawalIndex = preState.getNextWithdrawalIndex();
     int validatorIndex = preState.getNextWithdrawalValidatorIndex().intValue();
     for (int i = 0;
-        i < validatorCount
-            && expectedWithdrawals.size() < specConfigCapella.getMaxWithdrawalsPerPayload();
+        i < validatorCount && expectedWithdrawals.size() < maxValidatorsPerPayload;
         i++) {
-      final Validator validator = preState.getValidators().get(validatorIndex);
-      final UInt64 balance = preState.getBalances().get(validatorIndex).get();
-      if (predicates.isFullyWithdrawableValidator(validator, balance, epoch)) {
-        expectedWithdrawals.add(
-            withdrawalSchema.create(
-                withdrawalIndex,
-                UInt64.valueOf(validatorIndex),
-                new Bytes20(validator.getWithdrawalCredentials().slice(12)),
-                balance));
-        withdrawalIndex = withdrawalIndex.increment();
-      } else if (predicates.isPartiallyWithdrawableValidator(validator, balance)) {
-        expectedWithdrawals.add(
-            withdrawalSchema.create(
-                withdrawalIndex,
-                UInt64.valueOf(validatorIndex),
-                new Bytes20(validator.getWithdrawalCredentials().slice(12)),
-                balance.minus(specConfig.getMaxEffectiveBalance())));
-        withdrawalIndex = withdrawalIndex.increment();
+
+      final Validator validator = validators.get(validatorIndex);
+      if (predicates.hasEth1WithdrawalCredential(validator)) {
```

Comment:
Invert this to `if (!hasEth1WithdrawalCredentials) { continue; }` ?

Reply:
can't continue, need to update the reference at the end of the loop...